### ISSUE TITLE
File name containing glob wildcards cannot be captured in the process output 

### ISSUE THREAD

**Issue by mes5k** [nextflow-io/nextflow#219]
I have the following process, which will fail with missing expected output if the file name in question contains characters such as '[', ']', ':' and possibly others.

```
    process gzipem {
        publishDir path: "results", mode: "copy", overwrite: true

        input:
        file(json) from json_files

        output:
        file("${json}.gz") into gzipped_files

        script:
        """
        gzip -c ${json} > ${json}.gz
        """
    }
````

**Comment by mes5k** [comment]
Just some more information here.  I see similar problems if I'm trying to create files in a channel like this:
```
genome_dirs
    .splitText()
    .map{ it.trim() }
    .map{ file(it) }
```

Where ```genome_dirs``` is just a text file with a fully qualified path on each line.  Here's the exception I see:

```
Sep-06 16:24:45.372 [Actor Thread 2] ERROR n.extension.DataflowExtensions - @unknown
java.util.regex.PatternSyntaxException: Invalid range near index 172
/bioinformatics/internal_projects/genome_editing/archetype_dump/PUBLIC/Microbial/Viruses/ssDNA_viruses/Geminiviridae/Begomovirus/Bean_golden_yellow_mosaic_virus-[Puerto_Rico-Japan]/D00201.1/D00201.1.annotations.json.gz
                                                                                                                                                                            ^
    at sun.nio.fs.Globs.toRegexPattern(Globs.java:135)
    at sun.nio.fs.Globs.toUnixRegexPattern(Globs.java:211)
    at sun.nio.fs.UnixFileSystem.getPathMatcher(UnixFileSystem.java:294)
    at nextflow.file.FileHelper.getPathMatcherFor(FileHelper.groovy:696)
    at nextflow.file.FileHelper.visitFiles(FileHelper.groovy:734)
    at nextflow.file.FileHelper$visitFiles$5.call(Unknown Source)
    at nextflow.Nextflow.fileNamePattern(Nextflow.groovy:107)
    at sun.reflect.GeneratedMethodAccessor23.invoke(Unknown Source)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:498)
    at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:93)
    at org.codehaus.groovy.runtime.callsite.StaticMetaMethodSite$StaticMetaMethodSiteNoUnwrapNoCoerce.invoke(StaticMetaMethodSite.java:151)
    at org.codehaus.groovy.runtime.callsite.StaticMetaMethodSite.callStatic(StaticMetaMethodSite.java:102)
    at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callStatic(AbstractCallSite.java:222)
    at nextflow.Nextflow.file(Nextflow.groovy:133)
    at nextflow.Nextflow.file(Nextflow.groovy)
    at nextflow.Nextflow$file.callStatic(Unknown Source)
    at _nf_script_7b96e2d6$_run_closure3.doCall(_nf_script_7b96e2d6:21)
```

A brief glance at the code seems to indicate that ```file(it)``` creates a regex with the filename which the square brackets screw up. 

I tried adding this line:
```
.map{ "${it}".replace("[","\\[").replace("]","\\]") }
```
to the channel and the exception goes away, but any of the files with brackets in them get turned into ```[]``` linked list objects instead of actual files.

**Comment by pditommaso** [comment]
Looking for a patch. A workaround for your second problem should be 

```
genome_dirs
    .splitText()
    .map{ it.trim() }
    .map{ it as Path }
```

**Comment by mes5k** [comment]
Thanks Paolo!  ```as Path``` allows me to read the initial file, which I then rename sensibly, so everything downstream works fine.  If a proper fix to ```file``` is too much, then perhaps a note about poorly named files in the documentation?

**Comment by pditommaso** [comment]
Related problem See also #194.

**Comment by pditommaso** [comment]
It fix has been included in version `0.22.0-RC2`. Now glob wildcards can be escaped by using a backslash character. Also `file`, `Channel.fromPath` and process output file declaration support and optional parameter `glob` that can be set to `false` to disable glob path matching. For example: 

```
process foo {
  output:
  file 'file[a-b].txt' glob false into x 

  '''
  touch file\\[a-b\\].txt
  '''
}

```
